sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala

```diff
@@ -2023,3 +2023,143 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
 
   override def prettyName: String = "var_samp"
 }
+
+case class GpuReplaceNullmask(
+    input: Expression, 
+    mask: Expression) extends GpuExpression with ShimExpression {
+
+  override def dataType: DataType = input.dataType
+  override def nullable: Boolean = mask.nullable
+  override def children: Seq[Expression] = Seq(input, mask)
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    val res = withResource(input.columnarEval(batch)) { inputColumn =>
+      withResource(mask.columnarEval(batch)) { maskColumn =>
+        withResource(maskColumn.getBase.isNotNull) { maskColumnNotNull =>
+          inputColumn.getBase.copyWithBooleanColumnAsValidity(maskColumnNotNull)
+        }
+      }
+    }
+    GpuColumnVector.from(res, dataType)
+  }
+}
+
+object CudfMaxMinBy {
+  val KEY_ORDERING: String = "_key_ordering"
+  val KEY_VALUE: String = "_key_value"
+}
+
+abstract class CudfMaxMinByAggregate(
+    orderingType: DataType,
+    valueType: DataType) extends CudfAggregate {
+
+  protected val sortOrder: Int => cudf.OrderByArg
+
+  protected val reductionAggregation: ReductionAggregation
+
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = col => {
+    val orderCol = col.getChildColumnView(0).copyToColumnVector()
```

Comment:
This code is not the latest, missing the handle for the all-nulls case.
```
  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = col => {
    val orderColView = col.getChildColumnView(1)
    if (orderColView.getNullCount == orderColView.getRowCount) { // all nulls
      GpuScalar.from(null, dataType)
    } else {
      val tmpTable = withResource(orderColView.copyToColumnVector())(
        orderCol => new cudf.Table(orderCol, col)
      )
      val sorted = withResource(tmpTable) { _ =>
        // columns in table [order, original struct]
        tmpTable.orderBy(sortOrder(0))
      }
      withResource(sorted) { _ =>
        sorted.getColumn(1).reduce(ReductionAggregation.nth(0, NullPolicy.INCLUDE))
      }
    }
  }
```

Reply:
thanks, updated